platforms/pomelo/src/sim_drivers/emb_display.rs

```diff
@@ -0,0 +1,360 @@
+// CommanderTask - This is the "driver server"
+//////////////////////////////////////////////////////////////////////////////
+
+/// This task is spawned by the call to [`SimDisplay::register`]. It is a single
+/// async function that will process requests, and periodically redraw the
+/// framebuffer.
+struct CommanderTask {
+    cmd: KConsumer<Message<EmbDisplayService>>,
+    width: u32,
+    height: u32,
+}
+
+struct Context {
+    display: SimulatorDisplay<Gray8>,
+    framebuf: HeapArray<u8>,
+    width: u32,
+    height: u32,
+}
+
+impl CommanderTask {
+    /// The entrypoint for the driver execution
+    async fn run(self, width: u32, height: u32) {
+        let output_settings = OutputSettingsBuilder::new()
+            .scale(1)
+            .pixel_spacing(1)
+            .build();
+
+        let bytes = (width * height) as usize;
+
+        let display = SimulatorDisplay::<Gray8>::new(
+            (width, height),
+            &output_settings,
+            Some(graphics_container().as_ref()),
+        );
+        let framebuf = HeapArray::new(bytes, 0x00).await;
+        let context = Context {
+            display,
+            framebuf,
+            width,
+            height,
+        };
+        self.message_loop(context).await;
+    }
+
+    /// This loop services incoming client requests.
+    ///
+    /// Generally, don't handle errors when replying to clients, this indicates that they
+    /// sent us a message and "hung up" without waiting for a response.
+    async fn message_loop(&self, context: Context) {
+        debug!("display message loop");
+
+        let context = Rc::new(RefCell::new(context));
+
+        let (tx, rx): (
+            std::sync::mpsc::Sender<(
+                Rc<RefCell<Context>>,
+                OpenEnvelope<Result<Response, FrameError>>,
+                ReplyTo<EmbDisplayService>,
+                MonoChunk,
+            )>,
+            std::sync::mpsc::Receiver<(
+                Rc<RefCell<Context>>,
+                OpenEnvelope<Result<Response, FrameError>>,
+                ReplyTo<EmbDisplayService>,
+                MonoChunk,
+            )>,
+        ) = std::sync::mpsc::channel();
+
+        let rx = Rc::new(RefCell::new(rx));
+
+        let leakme = Closure::<dyn FnMut()>::new({
+            let rx = rx.clone();
+            move || {
+                let rx = rx.borrow_mut();
+                while let Ok((context, env, reply_tx, fc)) = rx.try_recv() {
+                    if let Ok(_) = draw_mono(&fc, &mut *context.borrow_mut()) {
+                        let response = env.fill(Ok(Response::DrawComplete(fc.into())));
+                        spawn_local(async {
+                            // TODO conflict with sleep logic
+                            // https://github.com/tosc-rs/mnemos/issues/256
+                            reply_tx.reply_konly(response).await.unwrap();
+                        });
```

Comment:
similarly i think having to spawn a whole task just to reply to a channel is pretty sad, it would be nice if we could figure out a way to just have the callback wake the async code, so that it could just keep doing async things...